plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+						FU_IO_CHANNEL_FLAG_FLUSH_INPUT, error)) {
+			g_prefix_error (error, "Failed to write command: ");
+			return FALSE;
+		}
+	}
+
+	for (int i = 0; i < MAX_RECV_MESSAGES; i++) {
+		g_autoptr(GBytes) rsp_bytes = NULL;
+		g_autoptr(XbSilo) silo = NULL;
+		g_autoptr(XbNode) response_node = NULL;
+
+		rsp_bytes = fu_io_channel_read_bytes (self->io_channel, -1, DEFAULT_RECV_TIMEOUT_MS, FU_IO_CHANNEL_FLAG_SINGLE_SHOT, error);
+		if (rsp_bytes == NULL) {
+			g_prefix_error (error, "Failed to read XML message: ");
+			return FALSE;
+		}
+
+		fu_firehose_updater_log_message ("reading", rsp_bytes);
+		if (!fu_firehose_updater_process_response (rsp_bytes, &silo, &response_node, error)) {
+			g_prefix_error (error, "Failed to parse XML message: ");
+			return FALSE;
+		}
+
+		if (silo && response_node) {
+			*out_silo = g_steal_pointer (&silo);
+			*out_response_node = g_steal_pointer (&response_node);
+			return TRUE;
+		}
+
+		/* continue until we get a 'response_node' */
+	}
+
+	g_set_error (error, G_IO_ERROR, G_IO_ERROR_TIMED_OUT,
+		     "Didn't get any response in the last %d messages", MAX_RECV_MESSAGES);
+	return FALSE;
+}
+
+static gboolean
+fu_firehose_updater_initialize (FuFirehoseUpdater *self, GError **error)
+{
+	guint n_msg = 0;
+
+	for (int i = 0; i < MAX_RECV_MESSAGES; i++) {
+		g_autoptr(GBytes) rsp_bytes = NULL;
+
+		rsp_bytes = fu_io_channel_read_bytes (self->io_channel, -1,
+						      (i == 0 ? INITIALIZE_INITIAL_TIMEOUT_MS : INITIALIZE_TIMEOUT_MS),
+						      FU_IO_CHANNEL_FLAG_SINGLE_SHOT, NULL);
+		if (rsp_bytes == NULL)
+			break;
+
+		fu_firehose_updater_log_message ("reading", rsp_bytes);
+		if (!fu_firehose_updater_process_response (rsp_bytes, NULL, NULL, error)) {
+			g_prefix_error (error, "Failed to parse XML message: ");
+			return FALSE;
+		}
+
+		n_msg++;
+	}
+
+	if (n_msg == 0) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+			     "Couldn't read initial firehose messages from device");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static guint
+fu_firehose_updater_configure (FuFirehoseUpdater *self, GError **error)
+{
+	gint max_payload_size = CONFIGURE_MAX_PAYLOAD_SIZE_TO_TARGET_IN_BYTES;
+
+	for (int i = 0; i < MAX_CONFIGURE_ATTEMPTS; i++) {
+		gchar *cmd_str;
+		GByteArray *cmd_bytearray = NULL;
+		g_autoptr(XbSilo) rsp_silo = NULL;
+		g_autoptr(XbNode) rsp_node = NULL;
+
+		cmd_str = g_strdup_printf ("<configure"
+					   " MemoryName=\"%s\""
+					   " Verbose=\"%d\""
+					   " AlwaysValidate=\"%d\""
+					   " MaxDigestTableSizeInBytes=\"%d\""
+					   " MaxPayloadSizeToTargetInBytes=\"%d\""
+					   " ZlpAwareHost=\"%d\""
+					   " SkipStorageInit=\"%d\""
+					   "/>",
+					   CONFIGURE_MEMORY_NAME,
+					   CONFIGURE_VERBOSE,
+					   CONFIGURE_ALWAYS_VALIDATE,
+					   CONFIGURE_MAX_DIGEST_TABLE_SIZE_IN_BYTES,
+					   max_payload_size,
+					   CONFIGURE_ZLP_AWARE_HOST,
+					   CONFIGURE_SKIP_STORAGE_INIT);
+		cmd_bytearray = g_byte_array_new_take ((guint8 *)cmd_str, strlen (cmd_str));
```

Comment:
would g_string_free_to_bytes be less clunky to use here? Then you can also append each attr in the element,rather than a super long list of args.

Reply:
switched to GString